zerver/migrations/0699_fix_user_role_system_groups.py

```diff
@@ -0,0 +1,191 @@
+        """
+        ROLE_COUNT_HUMANS = "11"
+        ROLE_COUNT_BOTS = "12"
+
+        human_counts = {
+            str(ROLE_REALM_ADMINISTRATOR): 0,
+            str(ROLE_REALM_OWNER): 0,
+            str(ROLE_MODERATOR): 0,
+            str(ROLE_MEMBER): 0,
+            str(ROLE_GUEST): 0,
+        }
+        for value_dict in (
+            UserProfile.objects.filter(realm=realm, is_bot=False, is_active=True)
+            .values("role")
+            .annotate(Count("role"))
+        ):
+            human_counts[str(value_dict["role"])] = value_dict["role__count"]
+        bot_count = UserProfile.objects.filter(realm=realm, is_bot=True, is_active=True).count()
+        return {
+            ROLE_COUNT_HUMANS: human_counts,
+            ROLE_COUNT_BOTS: bot_count,
+        }
+
+    def fix_user_memberships_if_needed(user_profile: Any) -> None:
+        realm = user_profile.realm
+
+        guest_group = get_system_user_group_by_name(SystemGroups.EVERYONE, realm.id)
+        member_group = get_system_user_group_by_name(SystemGroups.MEMBERS, realm.id)
+        moderator_group = get_system_user_group_by_name(SystemGroups.MODERATORS, realm.id)
+        admin_group = get_system_user_group_by_name(SystemGroups.ADMINISTRATORS, realm.id)
+        owner_group = get_system_user_group_by_name(SystemGroups.OWNERS, realm.id)
+
+        group_id_to_group_name = {
+            guest_group.id: "Guests",
+            member_group.id: "Members",
+            moderator_group.id: "Moderators",
+            admin_group.id: "Administrators",
+            owner_group.id: "Owners",
+        }
+        role_to_group = {
+            ROLE_GUEST: guest_group,
+            ROLE_MEMBER: member_group,
+            ROLE_MODERATOR: moderator_group,
+            ROLE_REALM_ADMINISTRATOR: admin_group,
+            ROLE_REALM_OWNER: owner_group,
+        }
+        group_id_to_role = {v.id: k for k, v in role_to_group.items()}
+
+        role_group_ids = [
+            guest_group.id,
+            member_group.id,
+            moderator_group.id,
+            admin_group.id,
+            owner_group.id,
+        ]
+
+        role = user_profile.role
+        group_implied_by_role = role_to_group[role]
+
+        direct_memberships = UserGroupMembership.objects.filter(
+            user_profile=user_profile, user_group_id__in=role_group_ids
+        )
+        if len(direct_memberships) == 0:
+            print(
+                f"User {user_profile.id} has no role group memberships. This is unexpected. Skipping."
+            )
+            return
+
+        if len(direct_memberships) > 1:
+            group_names = [group_id_to_group_name[m.user_group_id] for m in direct_memberships]
+            print(
+                f"User {user_profile.id} has more than one role group membership: {group_names}. "
+                f"Expected group based on role value of {role}: {group_id_to_group_name[group_implied_by_role.id]}. Skipping."
```

Comment:
Here also we can remove the user from incorrect user groups, instead of skipping.

Reply:
How are we handling "Full members" group memberships here? This case of being member of two system groups can be possible if user was a full member before their role being changed.

Reply:
The idea was ignoring them. full members groups was intentionally excluded from the system groups we fetch for the user, exactly to not trip up on this additional membership here. And I was counting on `promote_new_full_memberships` cronjob sorting this out later, by deleting the excess membership if needed. 

But now I found that it actually does `.exclude(waiting_period_threshold=0)`, so it won't clean this up for such realm. So added fetching and explicit cleanup now:
```
        if role != ROLE_MEMBER and fullmembers_membership is not None:
            memberships_to_delete.append(fullmembers_membership)
```